### ISSUE TITLE
tunable/grad traits

### ISSUE THREAD

**Issue by Mitchelldscott** [Mitchelldscott/control-rs#10]
## Summary

Like MATLabs tunable PID and SS objects it would be nice to have a universal tunable/grad trait to make any model format trainable.

## Proposal

1. Start with a generic variable type that works similar to nalgebra::Complex<T> and provides basic math
2. Add a random variable version that can create stochastic models with noise (with specified distribution and stationary/nonstationary)
3. add a trait for model formats to compute gradients that can be used to train variables (possibly only train variables with a trainable variant)

## ChatGPT implementation

* trainable should be an enum variant

```rust
use std::ops::{Add, Mul, Sub};

/// A generic tunable variable that supports basic math operations.
#[derive(Debug, Clone, Copy)]
pub struct Tunable<T> {
    value: T,
    trainable: bool,
}

impl<T> Tunable<T> {
    pub fn new(value: T, trainable: bool) -> Self {
        Self { value, trainable }
    }

    pub fn value(&self) -> &T {
        &self.value
    }

    pub fn is_trainable(&self) -> bool {
        self.trainable
    }
}

/// Implement basic math operations for `Tunable<T>` (assuming `T: Copy + Add<Output = T>`, etc.)
impl<T: Copy + Add<Output = T>> Add for Tunable<T> {
    type Output = Self;
    fn add(self, rhs: Self) -> Self::Output {
        Self::new(self.value + rhs.value, self.trainable || rhs.trainable)
    }
}

impl<T: Copy + Sub<Output = T>> Sub for Tunable<T> {
    type Output = Self;
    fn sub(self, rhs: Self) -> Self::Output {
        Self::new(self.value - rhs.value, self.trainable || rhs.trainable)
    }
}

impl<T: Copy + Mul<Output = T>> Mul for Tunable<T> {
    type Output = Self;
    fn mul(self, rhs: Self) -> Self::Output {
        Self::new(self.value * rhs.value, self.trainable || rhs.trainable)
    }
}

/// A trait for trainable models that can compute gradients.
pub trait TrainableModel {
    type Variable;

    /// Compute the gradient of the model with respect to the tunable variables
    fn compute_gradient(&self, x: f64, error: f64) -> Vec<Self::Variable>;

    /// Apply a gradient update (e.g., for optimization)
    fn apply_gradient(&mut self, gradients: &[Self::Variable], learning_rate: f64);
}

/// A tunable polynomial model: y = a_n * x^n + ... + a_1 * x + a_0
#[derive(Debug)]
pub struct PolynomialModel {
    coefficients: Vec<Tunable<f64>>, // Trainable polynomial coefficients
}

impl PolynomialModel {
    pub fn new(coeffs: Vec<f64>, trainable: bool) -> Self {
        Self {
            coefficients: coeffs.into_iter().map(|c| Tunable::new(c, trainable)).collect(),
        }
    }

    /// Evaluate the polynomial for a given x
    pub fn evaluate(&self, x: f64) -> f64 {
        self.coefficients
            .iter()
            .enumerate()
            .map(|(i, coef)| coef.value * x.powi(i as i32))
            .sum()
    }
}

impl TrainableModel for PolynomialModel {
    type Variable = Tunable<f64>;

    fn compute_gradient(&self, x: f64, error: f64) -> Vec<Self::Variable> {
        self.coefficients
            .iter()
            .enumerate()
            .map(|(i, coef)| {
                let grad = error * x.powi(i as i32); // Gradient of error w.r.t. coefficient
                Tunable::new(grad, coef.is_trainable())
            })
            .collect()
    }

    fn apply_gradient(&mut self, gradients: &[Self::Variable], learning_rate: f64) {
        for (coef, grad) in self.coefficients.iter_mut().zip(gradients.iter()) {
            if coef.is_trainable() {
                coef.value = coef.value - learning_rate * grad.value;
            }
        }
    }
}

fn main() {
    let mut poly = PolynomialModel::new(vec![1.0, -2.0, 3.0], true); // y = 3x² - 2x + 1
    println!("Initial Polynomial: {:?}", poly);

    let x = 2.0;
    let y_target = 10.0; // Desired output
    let y_pred = poly.evaluate(x);
    let error = y_pred - y_target; // Error signal

    let gradients = poly.compute_gradient(x, error);
    poly.apply_gradient(&gradients, 0.1);

    println!("Updated Polynomial: {:?}", poly);
}
```
